File: app/src/main/java/io/neft/Renderer/Item.java
package io.neft.Renderer;

import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.RectF;

import java.util.ArrayList;

import io.neft.Utils.RectFUtils;

public class Item {
    static void register(Renderer renderer) {
        renderer.actions.put(Renderer.InAction.CREATE_ITEM, new Action() {
            @Override
            void work(Reader reader) {
                new Item(reader.renderer);
            }
        });

        renderer.actions.put(Renderer.InAction.SET_ITEM_PARENT, new Action() {
            @Override
            void work(Reader reader) {
                reader.getItem().setParent(reader.getItem());
            }
        });

        renderer.actions.put(Renderer.InAction.INSERT_ITEM_BEFORE, new Action() {
            @Override
            void work(Reader reader) {
                reader.getItem().insertBefore(reader.getItem());
            }
        });

        renderer.actions.put(Renderer.InAction.SET_ITEM_VISIBLE, new Action() {
            @Override
            void work(Reader reader) {
                reader.getItem().setVisible(reader.getBoolean());
            }
        });

        renderer.actions.put(Renderer.InAction.SET_ITEM_CLIP, new Action() {
            @Override
            void work(Reader reader) {
                reader.getItem().setClip(reader.getBoolean());
            }
        });

        renderer.actions.put(Renderer.InAction.SET_ITEM_WIDTH, new Action() {
            @Override
            void work(Reader reader) {
                reader.getItem().setWidth(reader.getFloat());
            }
        });

        renderer.actions.put(Renderer.InAction.SET_ITEM_HEIGHT, new Action() {
            @Override
            void work(Reader reader) {
                reader.getItem().setHeight(reader.getFloat());
            }
        });

        renderer.actions.put(Renderer.InAction.SET_ITEM_X, new Action() {
            @Override
            void work(Reader reader) {
                reader.getItem().setX(reader.getFloat());
            }
        });

        renderer.actions.put(Renderer.InAction.SET_ITEM_Y, new Action() {
            @Override
            void work(Reader reader) {
                reader.getItem().setY(reader.getFloat());
            }
        });

        renderer.actions.put(Renderer.InAction.SET_ITEM_Z, new Action() {
            @Override
            void work(Reader reader) {
                reader.getItem().setZ(reader.getInteger());
            }
        });

        renderer.actions.put(Renderer.InAction.SET_ITEM_SCALE, new Action() {
            @Override
            void work(Reader reader) {
                reader.getItem().setScale(reader.getFloat());
            }
        });

        renderer.actions.put(Renderer.InAction.SET_ITEM_ROTATION, new Action() {
            @Override
            void work(Reader reader) {
                reader.getItem().setRotation(reader.getFloat());
            }
        });

        renderer.actions.put(Renderer.InAction.SET_ITEM_OPACITY, new Action() {
            @Override
            void work(Reader reader) {
                reader.getItem().setOpacity(reader.getInteger());
            }
        });

        renderer.actions.put(Renderer.InAction.SET_ITEM_BACKGROUND, new Action() {
            @Override
            void work(Reader reader) {
                reader.getItem().setBackground(reader.getItem());
            }
        });

        renderer.actions.put(Renderer.InAction.SET_ITEM_KEYS_FOCUS, new Action() {
            @Override
            void work(Reader reader) {
                reader.getItem().setKeysFocus(reader.getBoolean());
            }
        });
    }

    protected final Renderer renderer;

    public final int id;
    public float x = 0;
    public float y = 0;
    public float width = 0;
    public float height = 0;
    public float scale = 1f;
    public float rotation = 0f;
    public int opacity = 255;
    public int zIndex = 0;
    public boolean visible = true;
    public boolean clip = false;
    public Item parent;
    public Item background;
    public final ArrayList<Item> children;

    public final Matrix matrix = new Matrix();
    protected final float[] matrixValues = new float[]{0, 0, 0, 0, 0, 0, 0, 0, 1};
    protected boolean dirtyMatrix = false;

    protected boolean dirty = false;
    protected boolean dirtyChildren = false;
    public final RectF bounds = new RectF();
    public final RectF globalBounds = new RectF();
    protected final RectF oldGlobalBounds = new RectF();
    protected final Matrix globalMatrix = new Matrix();
    protected final RectF redrawRect = new RectF();

    private static final float PI = (float) Math.PI;

    public Item(Renderer renderer) {
        this.renderer = renderer;
        this.id = renderer.items.size();
        renderer.items.add(this);

        children = new ArrayList<>();
    }

    protected void invalidate() {
        if (this.dirty){
            return;
        }
        this.dirty = true;
        Item parent = this.parent;
        while (parent != null && !parent.dirtyChildren){
            parent.dirtyChildren = true;
            parent = parent.parent;
        }
    }

    protected void updateMatrix() {
        float a = 1;
        float b = 0;
        float c = 0;
        float d = 1;
        float tx = 0;
        float ty = 0;

        final float originX = width / 2;
        final float originY = height / 2;

        // translate to the origin
        tx = x + originX;
        ty = y + originY;

        // scale
        a = scale;
        d = scale;

        // rotate
        if (rotation != 0) {
            final float sr = (float) Math.sin(rotation);
            final float cr = (float) Math.cos(rotation);

            final float ac = a;
            final float dc = d;

            a = ac * cr;
            b = dc * sr;
            c = ac * -sr;
            d = dc * cr;
        }

        // translate to the position
        tx += a * -originX + c * -originY;
        ty += b * -originX + d * -originY;

        // save
        matrixValues[0] = a;
        matrixValues[1] = c;
        matrixValues[2] = tx;
        matrixValues[3] = b;
        matrixValues[4] = d;
        matrixValues[5] = ty;
        matrix.setValues(matrixValues);
    }

    public void setParent(Item val) {
        if (parent != null){
            parent.children.remove(this);
            parent.invalidate();
        }

        if (val != null){
            val.children.add(this);
            val.invalidate();
        }

        this.parent = val;
        invalidate();
    }

    public void insertBefore(Item val) {
        if (parent != null){
            parent.children.remove(this);
            parent.invalidate();
        }

        final Item parent = val.parent;
        final int index = parent.children.indexOf(val);
        parent.children.add(index, this);
        parent.invalidate();
        this.parent = parent;
        invalidate();
    }

    public void setVisible(boolean val) {
        visible = val;
        invalidate();
    }

    public void setClip(boolean val) {
        clip = val;
        invalidate();
    }

    public void setWidth(float val) {
        width = renderer.dpToPx(val);
        bounds.right = width;
        dirtyMatrix = true;
        invalidate();
    }

    public void setHeight(float val) {
        height = renderer.dpToPx(val);
        bounds.bottom = height;
        dirtyMatrix = true;
        invalidate();
    }

    public void setX(float val) {
        x = renderer.dpToPx(val);
        dirtyMatrix = true;
        invalidate();
    }

    public void setY(float val) {
        y = renderer.dpToPx(val);
        dirtyMatrix = true;
        invalidate();
    }

    public void setZ(int val) {
        zIndex = val;
        // TODO
    }

    public void setScale(float val) {
        scale = val;
        dirtyMatrix = true;
        invalidate();
    }

    public void setRotation(float val) {
        rotation = val;// * 180 / PI;
        dirtyMatrix = true;
        invalidate();
    }

    public void setOpacity(int val) {
        opacity = val;
        invalidate();
    }

    public void setBackground(Item val) {
        background = val;
        invalidate();
    }

    public void setKeysFocus(boolean val) {

    }

    protected void measure(final Matrix globalMatrix, RectF viewRect, final ArrayList<RectF> dirtyRects, final boolean forceUpdateBounds) {
        final boolean isDirty = forceUpdateBounds || dirty || dirtyMatrix;

        // break on no changes
        if (!dirtyChildren && !isDirty) {
            return;
        }

        // update transform
        if (dirtyMatrix) {
            updateMatrix();
        }

        // include local transform
        this.globalMatrix.setConcat(globalMatrix, matrix);

        // update bounds
        if (isDirty) {
            oldGlobalBounds.set(globalBounds);
            globalBounds.set(bounds);
            this.globalMatrix.mapRect(globalBounds);

            // add rectangle to redraw
            if (dirty || dirtyMatrix || !RectFUtils.equals(globalBounds, oldGlobalBounds)) {
                redrawRect.set(oldGlobalBounds);
                redrawRect.union(globalBounds);
                if (redrawRect.intersect(viewRect)) {
                    boolean redrawRectIncluded = false;
                    for (final RectF dirtyRect : dirtyRects) {
                        if (RectF.intersects(dirtyRect, redrawRect)) {
                            dirtyRect.union(redrawRect);
                            redrawRectIncluded = true;
                            break;
                        }
                    }
                    if (!redrawRectIncluded) {
                        dirtyRects.add(redrawRect);
                    }
                }

                dirty = false;
            }
        }

        // clip
        if (clip) {
            viewRect = globalBounds;
        }

        final boolean forceChildrenUpdate = forceUpdateBounds || dirtyMatrix;
        if (forceChildrenUpdate || dirtyChildren) {
            // measure background
            if (background != null) {
                background.measure(this.globalMatrix, viewRect, dirtyRects, forceChildrenUpdate);
            }

            // measure children
            for (final Item child : children) {
                child.measure(this.globalMatrix, viewRect, dirtyRects, forceChildrenUpdate);
            }

            // clear
            dirtyChildren = false;
            dirtyMatrix = false;
        }
    }

    protected void measure(final Matrix globalMatrix, final RectF viewRect, final ArrayList<RectF> dirtyRects) {
        measure(globalMatrix, viewRect, dirtyRects, false);
    }

    protected void drawShape(final Canvas canvas, final int alpha) {

    }

    protected void draw(final Canvas canvas, int alpha, final RectF rect) {
        if (opacity < 255) {
            alpha = Math.round(alpha * (opacity / 255f));
        }

        canvas.save();

        // transform
        canvas.concat(matrix);

        // clip
        if (clip){
            canvas.clipRect(0, 0, width, height);
        }

        // background
        if (background != null){
            background.draw(canvas, alpha, rect);
        }

        // shape
        if (RectF.intersects(rect, globalBounds)) {
            drawShape(canvas, alpha);
        }

        // render children
        for (final Item child : children){
            if (child.visible) {
                child.draw(canvas, alpha, rect);
            }
        }

        canvas.restore();
    }
}

File: Neft/Item.swift
import UIKit

extension Array where Element: Renderer.Object {
    func indexOfObject(object: Element) -> Int? {
        let length = self.count
        var index = 0
        while index < length {
            if self[index] === object {
                return index
            }
            index += 1
        }
        return nil
    }

    mutating func removeObject(object: Element) -> Bool {
        if let index = self.indexOfObject(object) {
            self.removeAtIndex(index)
            return true;
        }
        return false;
    }
}

class Item: Renderer.Object {
    class func register(app: GameViewController) {
        app.client.actions[InAction.CREATE_ITEM] = {
            (reader: Reader) in
            Item(app)
        }
        app.client.actions[InAction.SET_ITEM_PARENT] = {
            (reader: Reader) in
            (app.renderer.getObjectFromReader(reader) as! Item)
                .setParent(app.renderer.getObjectFromReader(reader) as? Item)
        }
        app.client.actions[InAction.INSERT_ITEM_BEFORE] = {
            (reader: Reader) in
            (app.renderer.getObjectFromReader(reader) as! Item)
                .insertBefore(app.renderer.getObjectFromReader(reader) as! Item)
        }
        app.client.actions[InAction.SET_ITEM_VISIBLE] = {
            (reader: Reader) in
            (app.renderer.getObjectFromReader(reader) as! Item)
                .setVisible(reader.getBoolean())
        }
        app.client.actions[InAction.SET_ITEM_CLIP] = {
            (reader: Reader) in
            (app.renderer.getObjectFromReader(reader) as! Item)
                .setClip(reader.getBoolean())
        }
        app.client.actions[InAction.SET_ITEM_WIDTH] = {
            (reader: Reader) in
            (app.renderer.getObjectFromReader(reader) as! Item)
                .setWidth(reader.getFloat())
        }
        app.client.actions[InAction.SET_ITEM_HEIGHT] = {
            (reader: Reader) in
            (app.renderer.getObjectFromReader(reader) as! Item)
                .setHeight(reader.getFloat())
        }
        app.client.actions[InAction.SET_ITEM_X] = {
            (reader: Reader) in
            (app.renderer.getObjectFromReader(reader) as! Item)
                .setX(reader.getFloat())
        }
        app.client.actions[InAction.SET_ITEM_Y] = {
            (reader: Reader) in
            (app.renderer.getObjectFromReader(reader) as! Item)
                .setY(reader.getFloat())
        }
        app.client.actions[InAction.SET_ITEM_SCALE] = {
            (reader: Reader) in
            (app.renderer.getObjectFromReader(reader) as! Item)
                .setScale(reader.getFloat())
        }
        app.client.actions[InAction.SET_ITEM_ROTATION] = {
            (reader: Reader) in
            (app.renderer.getObjectFromReader(reader) as! Item)
                .setRotation(reader.getFloat())
        }
        app.client.actions[InAction.SET_ITEM_OPACITY] = {
            (reader: Reader) in
            (app.renderer.getObjectFromReader(reader) as! Item)
                .setOpacity(reader.getInteger())
        }
        app.client.actions[InAction.SET_ITEM_BACKGROUND] = {
            (reader: Reader) in
            (app.renderer.getObjectFromReader(reader) as! Item)
                .setBackground(app.renderer.getObjectFromReader(reader) as? Item)
        }
        app.client.actions[InAction.SET_ITEM_KEYS_FOCUS] = {
            (reader: Reader) in
            (app.renderer.getObjectFromReader(reader) as! Item)
                .setKeysFocus(reader.getBoolean())
        }
    }

    var x: CGFloat = 0
    var y: CGFloat = 0
    var width: CGFloat = 0
    var height: CGFloat = 0
    var scale: CGFloat = 1
    var rotation: CGFloat = 0
    var opacity: CGFloat = 1
    var visible: Bool = true
    var clip: Bool = false
    var parent: Item?
    var background: Item?
    var children = [Item]()

    var transform = CGAffineTransform(a: 1, b: 0, c: 0, d: 1, tx: 0, ty: 0)

    internal private(set) var dirty = false
    internal var dirtyChildren = false
    internal var dirtyTransform = false
    var bounds = CGRect()
    var globalBounds = CGRect()

    internal func invalidate() {
        if dirty {
           return
        }
        dirty = true
        var parent = self.parent
        while parent != nil && !parent!.dirtyChildren {
            parent!.dirtyChildren = true
            parent = parent!.parent
        }
    }

    private func updateTransform() {
        var a: CGFloat = 1
        var b: CGFloat = 0
        var c: CGFloat = 0
        var d: CGFloat = 1
        var tx: CGFloat = 0
        var ty: CGFloat = 0

        let originX = width / 2
        let originY = height / 2

        // translate to the origin
        tx = x + originX
        ty = y + originY

        // scale
        a = scale
        d = scale

        // rotate
        if rotation != 0 {
            let sr = sin(rotation)
            let cr = cos(rotation)

            let ac = a
            let dc = d

            a = ac * cr
            b = dc * sr
            c = ac * -sr
            d = dc * cr
        }

        // translate to the position
        tx += a * -originX + c * -originY
        ty += b * -originX + d * -originY

        // save
        self.transform = CGAffineTransformMake(a, b, c, d, tx, ty)
    }

    internal func updateBounds() {
        bounds.size.width = width
        bounds.size.height = height
    }

    func setParent(val: Item?) {
        if parent != nil {
            parent!.children.removeObject(self)
            parent!.invalidate()
        }

        if val != nil {
            val!.children.append(self)
            val!.invalidate()
        }

        self.parent = val
        invalidate()
    }

    func insertBefore(val: Item) {
        if parent != nil {
            parent!.children.removeObject(self)
            parent!.invalidate()
        }

        let newParent = val.parent!
        let index = newParent.children.indexOfObject(val)!
        newParent.children.insert(self, atIndex: index)
        self.parent = newParent
        newParent.invalidate()
        invalidate()
    }

    func setVisible(val: Bool) {
        self.visible = val
        invalidate()
    }

    func setClip(val: Bool) {
        self.clip = val
        invalidate()
    }

    func setWidth(val: CGFloat) {
        self.width = val
        invalidate()
        dirtyTransform = true
        updateBounds()
    }

    func setHeight(val: CGFloat) {
        self.height = val
        invalidate()
        dirtyTransform = true
        updateBounds()
    }

    func setX(val: CGFloat) {
        self.x = val
        invalidate()
        dirtyTransform = true
    }

    func setY(val: CGFloat) {
        self.y = val
        invalidate()
        dirtyTransform = true
    }

    func setScale(val: CGFloat) {
        self.scale = val
        invalidate()
        dirtyTransform = true
    }

    func setRotation(val: CGFloat) {
        self.rotation = val
        invalidate()
        dirtyTransform = true
    }

    func setOpacity(val: Int) {
        self.opacity = CGFloat(val) / 255
        invalidate()
    }

    func setBackground(val: Item?) {
        self.background = val
        invalidate()
    }

    func setKeysFocus(val: Bool) {

    }

    func measure(var globalTransform: CGAffineTransform, var _ viewRect: CGRect, inout _ dirtyRects: [CGRect], forceUpdateBounds: Bool = false) {
        let isDirty = forceUpdateBounds || dirty || dirtyTransform

        // break on no changes
        if !dirtyChildren && !isDirty {
            return
        }

        // update transform
        if dirtyTransform {
            updateTransform()
        }

        // include local transform
        globalTransform = CGAffineTransformConcat(transform, globalTransform)

        // update bounds
        if isDirty {
            let oldGlobalBounds = globalBounds
            globalBounds = CGRectApplyAffineTransform(bounds, globalTransform)

            // add rectangle to redraw
            if dirty || dirtyTransform || !CGRectEqualToRect(globalBounds, oldGlobalBounds) {
                var redrawRect = CGRectUnion(oldGlobalBounds, globalBounds)
                redrawRect = CGRectIntersection(redrawRect, viewRect)
                if !CGRectIsEmpty(redrawRect) {
                    var i = 0
                    let length = dirtyRects.count
                    while i < length {
                        if CGRectIntersectsRect(dirtyRects[i], redrawRect) {
                            dirtyRects[i] = CGRectUnion(dirtyRects[i], redrawRect)
                            break
                        }
                        i += 1
                    }
                    if i == length {
                        dirtyRects.append(redrawRect)
                    }
                }

                dirty = false
            }
        }

        // clip
        if clip {
            viewRect = globalBounds
        }

        let forceChildrenUpdate = forceUpdateBounds || dirtyTransform
        if forceChildrenUpdate || dirtyChildren {
            // clear
            dirtyChildren = false
            dirtyTransform = false
            
            // measure background
            background?.measure(globalTransform, viewRect, &dirtyRects, forceUpdateBounds: forceChildrenUpdate)

            // measure children
            var i = 0
            let children = self.children
            let length = children.count
            while i < length {
                children[i].measure(globalTransform, viewRect, &dirtyRects, forceUpdateBounds: forceChildrenUpdate)
                i += 1
            }
        }
    }

    func drawShape(context: CGContextRef, inRect rect: CGRect) {

    }

    func draw(context: CGContextRef, inRect rect: CGRect) {
        CGContextSaveGState(context)

        // set opacity
        CGContextSetAlpha(context, opacity)

        // transform
        CGContextConcatCTM(context, transform)

        // clip
        if clip {
            CGContextClipToRect(context, bounds)
        }

        // background
        background?.draw(context, inRect: rect)

        // shape
        if CGRectIntersectsRect(rect, globalBounds) {
            drawShape(context, inRect: rect)
        }

        // render children
        for child in children {
            if child.visible {
                child.draw(context, inRect: rect)
            }
        }

        CGContextRestoreGState(context)
    }
}
